### ISSUE TITLE
Documentation structure proposal

### ISSUE THREAD

**Issue by joelgomes1994** [UPBGE/UPBGE-Docs#1]
Hello! Here I show my proposal on the documentation structure, I'm already working at it, but I'm publishing it to enable a discussion if it's really a good planning, or can be enhanced.

UPBGE Directed Manual
================

First, the documentation is a merge of [Blender manual (Game Engine section only)](https://docs.blender.org/manual/en/dev/game_engine/index.html) and [UPBGE's Python API (UPBGE and Standalone modules only)](https://pythonapi.upbge.org/).

The manual will only contain UPBGE's specific features or, at most, game dev specific tasks, and any generic definition will have a link to the related subject on the Blender official manual. For example, the manual may explain the game specific definitions on materials, but for general materials settings the manual will provide a link to Blender official manual, on the Materials section.

From now on, I will refer only to Blender manual pages inside [Game Engine section](https://docs.blender.org/manual/en/dev/game_engine/index.html).

Rework of Manual Tree Structure
======================

Current GE Manual
----------------------

The document's structure on official Blender manual is somewhat messy, with no contexts or sub themes, along with some pages that don't belong in there.

- Game Engine
   - Introduction
   - Game Logic Screen Layout
   - Game Settings
   - Game Materials
   - World
   - Logic
   - Camera
   - Physics
   - Performance Considerations
   - Python API
   - Standalone Player
   - Licensing of Blender Games

UPBGE Manual Proposal
---------------------------

My proposal is to make a context based manual:

**MANUAL**
- Introduction
   - Briefing
   - A Deeper Look
   - Game Logic Screen Layout
   - Compatibility Notes
- Editors
   - Properties Editor
      - Render
      - Scene
      - World
      - Object
      - Constraints
      - Texture
      - Physics
      - Materials
      - Data
   - Logic Editor
      - Properties Panel
      - Logic Bricks Area
      - Components Panel
- Logic _(Will be reworked with the release of logic nodes)_
   - Introduction
   - Sensors
   - Controllers
   - Actuators
   - Properties
   - States
   - Components
- Release Game
   - Licensing of Games
   - Performance Considerations
   - Standalone Player
   - Release Procedure

**API REFERENCE**
- UPBGE Modules (bge)
- Audio System (aud)
- OpenGL Wrapper (bgl)
- Font Drawing (blf)
- Math Types & Utilities (mathutils)

Note that some important topics aren't listed on the proposal (like Camera, World, Physics), it is because those topics will be part of Editors > Properties Editor, according to its subject (Camera in Data tab, World in World tab, and so on). Also, I thought about moving the Logic top section to Editors > Logic Editor's subsection, but it will depend on the upcoming logic nodes (Node Editor will be added to Editors).

I'll be waiting for enhancement comments on this proposal, thanks. :smile:

**Comment by UnidayStudio** [comment]
I loved the idea. It's very complete now. This initiative will attract lots of new users.

My suggestion, however, is to add a section on the beggining that guide the new users into their first "game" (not a game, just something beautiful moving on the screen). This will increse the retention of the game engine to new users. Because if in some short period of time, the new user creates something unique, the chances are that they will be less likely to abandon the software.

Something like the "Getting Started" tab in other game engines:
* Unreal Engine: https://docs.unrealengine.com/en-US/GettingStarted
* Unity: https://docs.unity3d.com/560/Documentation/Manual/GettingStarted.html
* Armory: https://armory3d.org/manual/#/

This section don't need to be complex. Just a shortcut to ensure that the new users will stay with us for a while. We can also create links between the getting started stuff and the real documentation stuff, like: "Un this area you can control the lamp settings, wich you can have a deep explanation later on, or by clicking here *link*".

I can write this section, if you guys want. :)

**Comment by panzergame** [comment]
Fantastic work ! Nice reorganization, indeed the previous documentation was not really maintained and even less reorganized.

About exposing mostly the difference from BFBGE and UPBGE you can link to this page https://doc.upbge.org/api-changes.php. 

I agree with @UnidayStudio about adding quick example, i have a short list of them:
- move a cube with a key fully from logic bricks
- move a cube from python script
- play an animation from brick
- play an animation from python
- add an object from brick
- add an object from python
- setup a python component

Thank you a lot for your work :)

**Comment by joelgomes1994** [comment]
@UnidayStudio @panzergame 
Thanks you both for the feedback, I just added a section called 'Tutorials' on the proposal top post, and 'getting started' tutorials will be inserted on them.

There will be a 'Basic Concepts' chapter on each tutorial, explaining concepts like 'pulses' (on logic bricks), the structure of Python classes and functions (on Python tutorial).

I will try to get some concepts from [Monster's Guides](https://blenderartists.org/t/monsters-resources/493045) on Blender Artists, his guides are brief and amazingly well made, I learned a lot from it on my early days of BGE.

**Comment by joelgomes1994** [comment]
Hey guys, could you take a look in [this page](https://upbge-docs.readthedocs.io/en/latest/manual/editors/properties/object.html)? It's somewhat the format I intend to continue writing the manual, exposing detailed info about UPBGE specific features, explaining Blender features focusing on UPBGE uses of these features while providing a link to Blender Manual for detailed info about these features. I want a feedback about what could be improved. Thanks. :smile:

**Comment by panzergame** [comment]
The page is correct, and descriptions are concise. Also what about linking API reference (e.g object color) in the manual ? Just say me if it is interesting or not.

Some minor notes, for future rewrite :
In https://upbge-docs.readthedocs.io/en/latest/manual/editors/properties/materials.html, The `alpha to coverage` mode allows to do clip alpha but using multiple sample (it works only when multi-samples is enable) and make smooth edges.

Also the billboard mode is using the Z object axis and not the Z world axis as before, which allow you to do any kind of billboards.

**Comment by joelgomes1994** [comment]
@panzergame Thanks for the answer.
The API links is a good idea indeed, not only for object color, but for any value available in GUI and Python.
About the Materials tab page, I didn't touched it yet. I've added some checkmarks on the structure proposal above showing which articles are done, which ones are present but need to be reworked (as the Materials one, from Blender manual) and which ones must be written.

**Comment by lordloki** [comment]
@joelgomes1994 I think that this structure can be enough to start the project. Good work!!! :-)
Once it is over we can think to integrate information from https://github.com/mikepan/GameEngineBook
I have created a wiki page to include the proposed structure.

P.D. I included a new "About this Manual" chapter.

**Comment by UnidayStudio** [comment]
@lordloki WOW! I never saw this book before. What an amazing book!

**Comment by joelgomes1994** [comment]
@lordloki I see that you have commited to this repository recently. When I was editing the docs, when trying to include the bpy API reference, I discovered that Read The Docs was not able to build it due to the large ammount of pages and overall size of it. It says "timeout" when building and don't finish it, needing the paid version of RTD. I don't know if you'll experience this issue, but is a migration to another domain viable (maybe Github Pages or something alike)?

**Comment by lordloki** [comment]
Yes, i rented weeks ago another server to update upbge.org webpage but i didn't have time to wake up yet. When is up we can add the bpy pages (i think that there is a limit of 400000 files).

**Comment by lordloki** [comment]
I've just updated the structure proposal. The new parts added we can fill them with Mike Pan game engine book and with @IzaZed Logic nodes documentation

**Comment by lordloki** [comment]
Just a remark. At this moment, I'm moving content from Mike pan book but i'm not adapting the content. I want to have all the content moved first and adapt this content after (rewording, new images, etc)

**Comment by lordloki** [comment]
Just for info about state:
- At this moment as the manual as the python api is on the upbge server (upbge.org/manual and upbge.org/api).
- I will update them once a week.

I like more the initial proposal to have both manual & api in a unique document but it is less work to have them indepent (I can almost automate the process).
Therefore, as long as we do not reach a good level I will leave the API independently.

Once reached we can see ways to join both documents.

**Comment by youle31** [comment]
Thanks very much for your works on documentation!

**Comment by pepsi112-dude** [comment]
The documentation is horrible, I cannot describe it.

Let's start shall we.
ImageToArray, you do not offer examples with it, you just say some stuff and that is it.


Parameters: | image – Image source object of type:  VideoFFmpeg ImageFFmpeg ImageBuff ImageMirror ImageMix ImageRender ImageViewport
-- | --


and a few other words, no tuts, how it connects to other things, how to take the array further and convert it to an image maybe. 
I searched for this array on google, I came up with a few pages and they are vague.

There is no tutorials on it how to use it in blender none. Upbge documentation has a few lines for it, this is not how a manual looks like.

Let's take from bge.texture class stuff like load, image placed there out of contex with no syntax attached to them.

load(imageBuffer, width, height), but how ?
But how there is no example, It just says load buffer from image, there is no explenation about it.
I had to search on other sites until my eyes popped out until I found something, I found that you can specify the source, to load to source such as imagebuff so you can replace the initiated buffer from texture.imagebuff, it gave me an error that with and height size do not match and I have not found anything regarding this neither on blender docs nor on the internet search engines. I just copied the error from the console and got zero results.

Error says buffer hasn't the correct size. and there is no refrence on anything where to search

This is even more disturbing:
 image
    Image data. (readonly)

Is it a plane, a rocket or superman, maybe it's a UFO because we don't know how to use it in context of a scripted line where this image is formulated like what Image= Magical mushrooms ? how do you use it , in what contex, with what. Syntax please. 

The results lack on google also because the lack of API refrence from blender game engine, the Bge documentation is the most "not there" on google. There are no tutorials, no real reffrence and example with sytax code. only Blender artists site where you have to beg for answers :-1: 

The documentation at least on bge module is mechanic, we are not robots.

**Comment by ShaunKulesa** [comment]
Pepsi112-dude, doing the docs is time consuming. Commenting on your "Examples" complaint, at the bottom of most pages it says examples. If you want some examples on there if you have any spare time you could make some and one of us will add them.

**Comment by lordloki** [comment]
Yeah, we know that the documentation lack a lot examples. We are trying to improve it but we have limited resources.
Regarding the other issues, for 0.2.5 it is not possible to use (and it never will) bpy for standalone. If you want to use it you will have to use 0.3 Alpha. Also, no new development effor will be put in 0.2.5 branch after 0.2.5b release (scheduled before year ends).

**Comment by pepsi112-dude** [comment]
> Yeah, we know that the documentation lack a lot examples. We are trying to improve it but we have limited resources.
> Regarding the other issues, for 0.2.5 it is not possible to use (and it never will) bpy for standalone. If you want to use it you will have to use 0.3 Alpha. Also, no new development effor will be put in 0.2.5 branch after 0.2.5b release (scheduled before year ends).

I have alpha 3, I was talking about 2.5 in the way that all the blender posts from the forum reagarding blender game engine is from that era.  I took a script from 2.5 and modified it, it's the rendertotexture script, beautiful projection with videotexture why it would be a shame not to be able to use blender.texture class and save images from it.

**Comment by pepsi112-dude** [comment]
> @pepsi112-dude , doing the docs is time consuming. Commenting on your "Examples" complaint, at the bottom of most pages it says examples. If you want some examples on there if you have any spare time you could make some and one of us will add them.

Do you think I have enough expiriance to write blender api code, I'm just a beginer in blender's api, it's consuming but so it is frustrating for others who search for a start on upbge  manual, who want to give blender game engine  a chance.

**Comment by ShaunKulesa** [comment]
@pepsi112-dude https://discord.gg/Db8DjnYS

**Comment by pepsi112-dude** [comment]
> @pepsi112-dude https://discord.gg/Db8DjnYS

Thanks Shaun, I will check this out

**Comment by rcampos0** [comment]
hello, everyone, a while ago, the group discussion, and the group quote about the Game Engine Book, I went to the repository, and forked the repository, for my github account. In order to translate it into Brazilian Portuguese, and start contributing something I can. you could take a look, a review has not been done, I’m just translated, there’s a little more to go until the end. https://github.com/rcampos0/GameEngineBook

**Comment by rpaladin** [comment]
Hi @joelgomes1994.
Are you still interested helping with documentation?
Currently images are still on legacy versions and not current.

Additionally, many of the links in your first post are outdated.

**Comment by aum7** [comment]
i would like to know the purpose of :
_UPBGE-Docs / drafts_ folder

where should i move files from : 
- _UPBGE-Docs / drafts / camera_ folder ? 
- _UPBGE-Docs / drafts / physics_ folder ?
- _UPBGE-Docs / drafts / videotexture.rst_ file ?
